### ISSUE TITLE
Breaking Time Machine on OS X

### ISSUE THREAD

**Issue by chbrown** [npm/npm#4664]
This may be a Time Machine bug, but npm is the only perpetrator of this oddity that I have dealt with.

Steps to reproduce:

1. Have a npm-published package somewhere on your filesystem `xyz`, and globally link it (e.g., for easier development) with `npm link`
2. Run `npm link xyz` in some other node.js package / folder
3. Backup your computer with Time Machine 
4. Run `npm install -g xyz` to install that `xyz` from the npm registry, e.g., if you forgot that it was locally linked.
5. Try to backup again...

Time Machine will balk with a useless error message, but Console.app displays somewhat more informative com.apple.backupd log lines, pointing to `/usr/local/lib/node_modules/xyz/*` files with the message: "Error: (-36) SrcErr:NO".

A simple `npm uninstall -g xyz` fixes all, but it's initially perplexing because Time Machine is so tight-lipped about the actual error.

Now that I think about it, step 2 is probably not required.

I presume the error is that Time Machine trips up when it hits a real folder where there was previously a symlink, even though the directory's contents are identical. It's really Time Machine's fault, but I'm surprised that this doesn't crop up with other programs. Is npm doing something weird with its symlinks?

**Comment by othiym23** [comment]
After (some time and) reading the post linked by @feross, I'm going to go ahead and say that this isn't npm's issue to fix. Time Machine does some gross stuff under the hood involving links and directories, and it doesn't surprise me that this case falls between the cracks a bit. The good news is that I _think_ this has been addressed in OS X / macOS / whatever we're calling it this week, the bad news is that even if we wanted to work around this, it would be beyond the capabilities of the CLI team to fix (given that Time Machine runs on a regular schedule, and disallowing switching `node_modules` between links and real installs would be so user-hostile that it would effectively break `npm link`). Thanks to all for your time, links, and patience.

**Comment by chbrown** [comment]
You're right that it's Time Machine's problem — but Time Machine doesn't have an issue tracker :(

Oddly enough, I haven't run into this problem for a while. I did set `tmutil addexclusion /usr/local/lib/node_modules` at one time but it doesn't seem to take, because `tmutil isexcluded /usr/local/lib/node_modules` says "[Included]". Maybe it got fixed somewhere between Mountain Lion (I was a late adopter) and El Capitan? I've also replaced some of my `npm link`-ing with [npm-reallink](https://www.npmjs.com/package/npm-reallink) (which I wrote to facilitate npm-linking local TypeScript projects), which only symlinks files, not directories.

Anyhow, thanks for following up — I agree that there's no good workaround that npm could/should implement, and partly meant for this issue just to be sort of a PSA to other npm + Time Machine users.
